packages/vite/src/client/client.ts

```diff
@@ -315,23 +315,48 @@ async function waitForSuccessfulPing(
 ) {
   const pingHostProtocol = socketProtocol === 'wss' ? 'https' : 'http'
 
-  // eslint-disable-next-line no-constant-condition
-  while (true) {
+  const ping = async () => {
+    // A fetch on a websocket URL will return a successful promise with status 400,
+    // but will reject a networking error.
+    // When running on middleware mode, it returns status 426, and an cors error happens if mode is not no-cors
     try {
-      // A fetch on a websocket URL will return a successful promise with status 400,
-      // but will reject a networking error.
-      // When running on middleware mode, it returns status 426, and an cors error happens if mode is not no-cors
       await fetch(`${pingHostProtocol}://${hostAndPath}`, {
         mode: 'no-cors',
       })
-      break
-    } catch (e) {
-      // wait ms before attempting to ping again
-      await new Promise((resolve) => setTimeout(resolve, ms))
+      return true
+    } catch {}
+    return false
+  }
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    if (document.visibilityState === 'visible') {
+      if (await ping()) {
+        break
+      }
+      await wait(ms)
+    } else {
+      await Promise.race([wait(ms), waitForWindowShow()])
```

Comment:
This will register a new event handler for `visibilitychange` every second while the visibility doesn't change (then resolve all of them at the same time. I think it would be good to create a single `waitForWindowShow` promise and use it here until it resolves.

Reply:
Good catch! I fixed this 👍

Reply:
I didn't notice this too, I thought it would wait for window show altogether before continuing the while loop, maybe we could so a simple `await waitForWindowShow()` here?

Reply:
Ah, you're right. I don't know why I wrote it like this 😅